File: core/middleware/trace.py
from starlette.types import ASGIApp, Scope, Receive, Send

from core.utils.trace_utils import generate_trace_id, set_trace_id, clear_trace_id, TRACE_ID_HEADER


class TraceMiddleware:
    """ASGI middleware that manages trace ID lifecycle per request.

    Sets a trace ID at the start of each request and clears it on completion.
    If the incoming request has a ``traceId`` header (e.g. from an upstream
    service), that value is propagated; otherwise a new UUID is generated.
    """

    def __init__(self, app: ASGIApp):
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        trace_id = ""
        for name, value in scope.get("headers", []):
            if name == TRACE_ID_HEADER.encode("latin-1"):
                trace_id = value.decode("latin-1")
                break

        if not trace_id:
            trace_id = generate_trace_id()

        set_trace_id(trace_id)

        try:
            await self.app(scope, receive, send)
        finally:
            clear_trace_id()

File: core/middleware/trace.go
package middleware

import (
	"hei-gin/core/utils"

	"github.com/gin-gonic/gin"
)

// Trace returns a Gin middleware that manages trace ID lifecycle per request.
//
// It reads a trace_id from the incoming request header. If the header is empty,
// a new trace ID is generated. The trace ID is stored in the Gin context under
// the key "trace_id" for downstream handlers and loggers to use.
func Trace() gin.HandlerFunc {
	return func(c *gin.Context) {
		traceID := c.GetHeader("trace_id")
		if traceID == "" {
			traceID = utils.GenerateTraceID()
		}
		c.Set("trace_id", traceID)

		c.Next()
	}
}
